File: src/01-easy/retro-typewriter-art/solution.ml
let ab = [("sp", ' '); ("bS", '\\'); ("sQ", '\''); ("nl", '\n')]

let decode tok =
  let len = String.length tok in
  let l2 = if len >= 2 then String.sub tok (len-2) 2 else "" in
  match List.assoc_opt l2 ab with
  | Some ch ->
    let num = String.sub tok 0 (len-2) in
    let n = if num = "" then 1 else max 1 (int_of_string num) in
    String.make n ch
  | None ->
    let ch = tok.[len-1] in
    let num = String.sub tok 0 (len-1) in
    let n = if num = "" then 1 else max 1 (int_of_string num) in
    String.make n ch

let () =
  let line = input_line stdin in
  let toks = String.split_on_char ' ' line in
  print_endline (String.concat "" (List.map decode toks))

File: src/01-easy/retro-typewriter-art/solution.lua
local ab = {sp=' ', bS='\\', sQ="'", nl='\n'}
local line = io.read()
local out = {}
for tok in line:gmatch('%S+') do
    local l2 = #tok >= 2 and tok:sub(-2) or ''
    local ch, num
    if ab[l2] then ch = ab[l2]; num = tok:sub(1, -3)
    else ch = tok:sub(-1); num = tok:sub(1, -2) end
    local n = (#num > 0) and tonumber(num) or 1
    if n < 1 then n = 1 end
    for i = 1, n do out[#out+1] = ch end
end
print(table.concat(out))
